components/gitpod-db/src/typeorm/user-db-impl.ts

```diff
@@ -350,6 +356,95 @@ export class TypeORMUserDBImpl implements UserDB {
+        log.info(`getUserByCredentials ${identifier}`)
+        const user = await this.findUserById(identifier);
+        if (user) {
+            log.info(`getUserByCredentials returned: ${user.id}, ${user.name}`)
+            return {
+                id: user.id,
+                name: user.name,
+            };
+        }
+        log.info(`getUserByCredentials returned nothing!`)
+        return;
+    }
+    public async extraAccessTokenFields?(user: OAuthUser): Promise<ExtraAccessTokenFields | undefined> {
+        // No extra fields in token
+        log.info(`getUserByCredentials ${JSON.stringify(user)}`);
+        return;
+    }
+
+    // OAuthTokenRepository
+    async issueToken(client: OAuthClient, scopes: OAuthScope[], user?: OAuthUser): Promise<OAuthToken> {
+        const expiry = tokenExpiryInFuture.getEndDate();
+        log.info(`issueToken ${JSON.stringify(expiry)}`)
+        return <OAuthToken>{
+            accessToken: crypto.randomBytes(30).toString('hex'),
+            accessTokenExpiresAt: expiry,
+            client,
+            user,
+            scopes: scopes,
+        };
+    }
+    async issueRefreshToken(accessToken: OAuthToken): Promise<OAuthToken> {
+        // NOTE(rl): this exists for the OAuth server code - Gitpod tokens are non-refreshable (atm)
+        accessToken.refreshToken = "refreshtokentoken";
+        accessToken.refreshTokenExpiresAt = new DateInterval("30d").getEndDate();
+        await this.persist(accessToken);
+        return accessToken;
+    }
+    async persist(accessToken: OAuthToken): Promise<void> {
+        log.info(`persist access token ${JSON.stringify(accessToken)}`);
+        var scopes: string[] = [];
+        for (const scope of accessToken.scopes) {
+            scopes = scopes.concat(scope.name);
+        }
+
+        // Does the token already exist?
+        var dbToken: GitpodToken & { user: DBUser };
+        const tokenHash = crypto.createHash('sha256').update(accessToken.accessToken, 'utf8').digest("hex");
+        const userAndToken = await this.findUserByGitpodToken(tokenHash);
+        if (userAndToken) {
+            // Yes, update it (~)
+            // NOTE(rl): as we don't support refresh tokens yet this is not really required 
+            // since the OAuth server lib calls issueRefreshToken immediately after issueToken
+            // We do not allow changes of name, type, user or scope.
+            dbToken = userAndToken.token as GitpodToken & { user: DBUser };
+            const repo = await this.getGitpodTokenRepo();
+            log.info(`persist access token update: ${JSON.stringify(dbToken)}`);
+            return repo.updateById(tokenHash, dbToken);
+        } else {
+            var user: MaybeUser;
+            if (accessToken.user) {
+                user = await this.findUserById(accessToken.user.id)
+            }
+            dbToken = {
+                tokenHash,
+                name: `local-app`,
```

Comment:
Would it make sense to include the name of the requesting application here? This way we could have multiple tokens issued using OAuth.

Reply:
Great point.

Hard-coding leftovers replaced with the client name.